### PR TITLE
Use chemical compositions in rheology models for parameter parsing

### PR DESCRIPTION
**Pull request by anne-glerum**
For all the rheology models that the visco_plastic material model depends on, this PR uses the MapParsing functionality to retrieve material properties from the input file. This way, the parameter file does not have to specify properties for fields that to not represent chemical compositions (e.g., plastic strain). This saves entering a lot of values when stresses and/or strains are tracked, whose field type is automatically deduced from their field name set in the input file.

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](https://aspect-documentation.readthedocs.io/en/latest/user/extending/testing/writing-tests.html) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

### CODE REVIEW

**Review by bobmyhill — CHANGES_REQUESTED** [review]
Brilliant, thanks @anne-glerum! A couple of very minor comments, but as long as the tests pass, this is basically ready to go.

File: source/material_model/visco_plastic.cc [hunk 1]
```diff
@@ -232,29 +232,29 @@ namespace aspect
               // TODO: This is only consistent with viscosity averaging if the arithmetic averaging
               // scheme is chosen. It would be useful to have a function to calculate isostress viscosities.
               isostrain_viscosities =
-                rheology->calculate_isostrain_viscosities(in, i, volume_fractions_for_rheology, phase_function_values, phase_function.n_phase_transitions_for_each_composition());
```
[the remark this answers is not in the document]
Comment (anne-glerum) [reply source/material_model/visco_plastic.cc:235]: outdated

File: source/material_model/rheology/friction_models.cc [hunk 2]
```diff
@@ -243,6 +249,8 @@ namespace aspect
 
         // Get the number of fields for composition-dependent material properties
         // including the background field.
+        // TODO Make sure functions only have to be specified per chemical composition,
```
[the remark this answers is not in the document]
Comment (anne-glerum) [reply source/material_model/rheology/friction_models.cc:252]: Looking into it more, I think it should be adapted in this PR. The index used to select the right function is based on looping over the size of the chemical composition only volume_fractions. It is passed on by rheology/visco_plastic to the compute_friction_angle function. That function has functions for _all_ compositions, so then the index would be wrong.

File: source/material_model/visco_plastic.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (bobmyhill) [reply source/material_model/visco_plastic.cc:235]: Sorry, this wasn't very clear. I meant I think lines 182-184 can now be removed.

[the remark this answers is not in the document]
Comment (bobmyhill) [reply source/material_model/visco_plastic.cc:235]: Thanks :)

File: source/material_model/diffusion_dislocation.cc [hunk 3]
```diff
@@ -298,13 +299,13 @@ namespace aspect
           prm.declare_entry ("Densities", "3300.",
                              Patterns::List(Patterns::Double(0.)),
                              "List of densities, $\\rho$, for background mantle and compositional fields, "
-                             "for a total of N+1 values, where N is the number of compositional fields. "
+                             "for a total of N+1 values, where N is the number of all or only all chemical compositional fields. "
```
Comment (bobmyhill) [inline source/material_model/diffusion_dislocation.cc:302]:
```suggestion
                             "for a total of N+1 values, where N is the number of all compositional fields or only those corresponding to chemical compositions. "
```

File: source/material_model/diffusion_dislocation.cc [hunk 4]
```diff
@@ -298,13 +299,13 @@ namespace aspect
...
                              "If only one value is given, then all use the same value. "
                              "Units: \\si{\\kilogram\\per\\meter\\cubed}.");
           prm.declare_entry ("Thermal expansivities", "3.5e-5",
                              Patterns::List(Patterns::Double(0.)),
                              "List of thermal expansivities for background mantle and compositional fields, "
-                             "for a total of N+1 values, where N is the number of compositional fields. "
+                             "for a total of N+1 values, where N is the number of all or only all chemical compositional fields. "
```
Comment (bobmyhill) [inline source/material_model/diffusion_dislocation.cc:308]: ditto here and in all the other files. Sorry! Hopefully you can just find and replace.

**Comment by anne-glerum** [comment]
Thanks for the quick review @bobmyhill, I've addressed your comments and the testsers pass (except for gmg tests on dealii-master).

**Review by bobmyhill — APPROVED** [review]

**Comment by bobmyhill** [comment]
Brilliant, thanks! This will make prms so much neater.
Is this ready to merge from your end? I'll give it one last check to double check all the rheology function calls, and will then wait for your ok to merge.

File: doc/modules/changes/20240122_glerum [hunk 5]
```diff
@@ -0,0 +1,6 @@
+ Improved: For the visco_plastic and visco_elastic
```
Comment (bobmyhill) [inline doc/modules/changes/20240122_glerum:1]:
```suggestion
 Improved: For the diffusion_dislocation, visco_plastic and visco_elastic
```

**Comment by anne-glerum** [comment]
Yes, from my side this is ready, thank you.

**Comment by bobmyhill** [comment]
Ok, grand! Last thing; would you mind squashing a few commits? Perhaps the 4 test-related results (32be39b--795e312) and then 6190a41--c96986d could be squashed together.

**Pull request merged**